contracts/contracts/coordination/InfractionCollector.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import "./Coordinator.sol";
+import "../../threshold/ITACoChildApplication.sol";
+
+contract InfractionCollector is OwnableUpgradeable {
+    Coordinator public coordinator;
+
+    // Reference to the TACoChildApplication contract
+    ITACoChildApplication public tacoChildApplication;
+
+    // Mapping to keep track of reported infractions
+    mapping(uint32 => mapping(address => bool)) public infractions;
+
+    function initialize(
+        Coordinator _coordinator,
+        ITACoChildApplication _tacoChildApplication
+    ) public initializer {
+        __Ownable_init(msg.sender);
+        coordinator = _coordinator;
+        tacoChildApplication = _tacoChildApplication;
+    }
+
+    function reportMissingTranscript(
+        uint32 ritualId,
+        address[] calldata stakingProviders
+    ) external {
+        // Ritual must have failed
+        require(
+            coordinator.getRitualState(ritualId) == Coordinator.RitualState.DKG_TIMEOUT,
+            "Ritual must have failed"
+        );
+
+        for (uint256 i = 0; i < stakingProviders.length; i++) {
+            // Check if the infraction has already been reported
+            require(!infractions[ritualId][stakingProviders[i]], "Infraction already reported");
```

Comment:
Given that we are going through a list of addresses, should we use an `if` instead of a `require`?

Reply:
Unsure about this, I see benefits on both sides, but I'd lean towards reverting the TX. 
BTW, I think there should be an `InfractionCollection` specific event, something like `ReportedMissingTranscript` or similar.

Reply:
fixed in https://github.com/nucypher/nucypher-contracts/pull/267/commits/3bdcc8dcf613489042ff14aa9e3eb460e18a4270